cmd/devicesimulator/main.go

```diff
@@ -167,23 +165,35 @@ func reportVersion(versionFormat *string) error {
 
 func startAgent(ctx context.Context, agent *agent.Agent, log *logrus.Logger, agentInstance int) {
 	activeAgents.Inc()
-	prefix := agent.GetLogPrefix()
-	err := agent.Run(ctx)
-	if err != nil {
-		// agent timeout waiting for enrollment approval
-		if errors.Is(err, wait.ErrWaitTimeout) {
-			log.Errorf("%s: agent timed out: %v", prefix, err)
-		} else if ctx.Err() != nil {
-			// normal teardown
-			log.Infof("%s: agent stopped due to context cancellation.", prefix)
-		} else {
-			log.Fatalf("%s: %v", prefix, err)
+	err := wait.PollImmediateWithContext(ctx, 2*time.Second, 5*time.Minute, func(ctx context.Context) (bool, error) {
+		// timeout after 30s and retry
+		ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+		defer cancel()
+
+		prefix := agent.GetLogPrefix()
+		if err := agent.Run(ctx); err != nil {
+			// agent timeout waiting for enrollment approval
+			if errors.Is(err, wait.ErrWaitTimeout) {
+				log.Errorf("%s: agent %d timed out: %v", prefix, agentInstance, err)
+				return false, nil
+			} else if ctx.Err() != nil {
+				// normal teardown
+				log.Infof("%s: agent %d stopped due to context cancellation.", prefix, agentInstance)
+				return false, nil
+			} else {
+				log.Fatalf("%s: %v", prefix, err)
```

Comment:
If the agent hits a 500 error will we retry? Do we know why we received a 500 error?

Reply:
You mentioned in Jira card that this error is `just a warning`, so we need just to retry it. Changed this func to retry in any case. In the worst case it will fail after `wait.PollImmediateWithContext` with a `timed out waiting for the condition` error.